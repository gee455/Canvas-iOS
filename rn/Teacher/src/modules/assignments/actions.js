//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import { createAction } from 'redux-actions'
import canvas from '../../canvas-api'
import { UPDATE_COURSE_DETAILS_SELECTED_TAB_SELECTED_ROW_ACTION } from '../courses/actions'

export let AssignmentListActions = (api: CanvasApi): * => ({
  refreshAssignmentList: createAction('assignmentList.refresh', (courseID: string, gradingPeriodID?: string, includeSubmissions?: boolean = false) => {
    const include = ['assignments', 'all_dates', 'overrides', 'discussion_topic']
    if (includeSubmissions) {
      include.push('submission')
    }
    return {
      promise: api.getAssignmentGroups(courseID, gradingPeriodID, include),
      courseID,
      gradingPeriodID,
    }
  }),
  refreshAssignment: createAction('assignment.refresh', (courseID: string, assignmentID: string) => {
    return {
      promise: api.getAssignment(courseID, assignmentID),
      courseID,
      assignmentID,
    }
  }),
  refreshAssignmentDetails: createAction('assignment.refreshDetails', (courseID: string, assignmentID: string, getSubmissionStatus?: boolean) => {
    let promises = [api.getAssignment(courseID, assignmentID)]
    if (getSubmissionStatus) {
      promises.push(api.refreshSubmissionSummary(courseID, assignmentID))
    }

    return {
      promise: Promise.all(promises),
      courseID,
      assignmentID,
    }
  }),
  updateAssignment: createAction('assignment.update', (courseID, updatedAssignment, originalAssignment) => {
    return {
      promise: api.updateAssignment(courseID, updatedAssignment),
      courseID,
      updatedAssignment,
      originalAssignment,
      handlesError: true,
    }
  }),
  cancelAssignmentUpdate: createAction('assignment.cancel-update', (originalAssignment) => {
    return { originalAssignment }
  }),
  refreshGradeableStudents: createAction('assignment.gradeable-students', (courseID: string, assignmentID: string) => {
    return {
      promise: api.getAssignmentGradeableStudents(courseID, assignmentID),
      courseID,
      assignmentID,
    }
  }),
  updateCourseDetailsSelectedTabSelectedRow: createAction(UPDATE_COURSE_DETAILS_SELECTED_TAB_SELECTED_ROW_ACTION, (rowID: string) => {
    return { rowID }
  }),
  anonymousGrading: createAction('assignment.anonymous', (courseID: string, assignmentID: string, anonymous: boolean) => ({
    courseID, assignmentID, anonymous,
  })),
})

export default (AssignmentListActions(canvas): *)
