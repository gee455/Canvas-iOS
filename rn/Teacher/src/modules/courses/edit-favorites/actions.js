//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import { createAction } from 'redux-actions'
import canvas from '../../../canvas-api'

export let FavoritesActions = (api: CanvasApi): * => ({
  toggleCourseFavorite: createAction('courses.toggleFavorite', (courseID: string, markAsFavorite: boolean) => {
    return {
      promise: markAsFavorite ? api.favoriteCourse(courseID) : api.unfavoriteCourse(courseID),
      courseID,
      markAsFavorite,
      syncToNative: true,
    }
  }),
})

export default (FavoritesActions(canvas): *)
