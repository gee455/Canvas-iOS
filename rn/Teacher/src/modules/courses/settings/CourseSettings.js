//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/**
 * Created by bkraus on 3/8/17.
 * @flow
 */

import React, { Component } from 'react'
import { connect } from 'react-redux'
import {
  View,
  StyleSheet,
  TouchableHighlight,
  PickerIOS,
  Image,
  LayoutAnimation,
} from 'react-native'

import i18n from 'format-message'
import colors from '../../../common/colors'
import { mapStateToProps } from './map-state-to-props'
import CourseSettingsActions from './actions'
import ModalOverlay from '../../../common/components/ModalOverlay'
import { alertError } from '../../../redux/middleware/error-handler'
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view'
import { Text, TextInput } from '../../../common/text'
import Screen from '../../../routing/Screen'
import Navigator from '../../../routing/Navigator'
import branding from '../../../common/branding'

var PickerItemIOS = PickerIOS.Item

type Props = {
  navigator: Navigator,
  course: Course,
  color: string,
  updateCourse: (Course, Course) => Course,
  pending: number,
  error: ?string,
}

const DISPLAY_NAMES = new Map([
  ['feed', i18n('Course Activity Stream')],
  ['wiki', i18n('Pages Front Page')],
  ['modules', i18n('Course Modules')],
  ['assignments', i18n('Assignments List')],
  ['syllabus', i18n('Syllabus')],
])

export class CourseSettings extends Component<Props, any> {
  constructor (props: Props) {
    super(props)

    this.state = {
      name: this.props.course.original_name ? this.props.course.original_name : this.props.course.name,
      home: this.props.course.default_view,
      pending: false,
    }
  }

  course = () => ({
    ...this.props.course,
    name: this.state.name,
    default_view: this.state.home,
  })

  componentWillReceiveProps (props: Props) {
    if (props.error) {
      this.setState({ pending: false })

      setTimeout(() => {
        alertError(props.error)
      }, 100)
    }
    this.state.pending && !props.pending && !props.error && this.props.navigator.dismissAllModals()
  }

  done = () => {
    this.setState({ pending: true })
    this.props.updateCourse(this.course(), this.props.course)
  }

  _togglePicker = () => {
    let animation = LayoutAnimation.create(250, LayoutAnimation.Types.linear, LayoutAnimation.Properties.opacity)
    LayoutAnimation.configureNext(animation)
    this.setState({ showingPicker: !this.state.showingPicker })
  }

  render () {
    let pickerDetailStyle = this.state.showingPicker ? { color: branding.primaryBrandColor } : {}

    return (
      <Screen
        title={i18n('Course Settings')}
        drawUnderNavBar={false}
        navBarTitleColor={colors.darkText}
        navBarButtonColor={colors.link}
        rightBarButtons={[{
          style: 'done',
          testID: 'course-settings.done-btn',
          title: i18n('Done'),
          action: this.done,
        }]}
        dismissButtonTitle={i18n('Cancel')}
      >
        <View style={{ flex: 1 }}>
          <ModalOverlay text={i18n('Saving')} visible={this.state.pending} />
          <KeyboardAwareScrollView
            style={styles.scrollView}>
            <View style={styles.header}>
              <View style={styles.headerContent}>
                {Boolean(this.props.course.image_download_url) &&
                    <Image source={{ uri: this.props.course.image_download_url }} style={styles.headerContent} />
                }
                <View style={[styles.headerContent, { backgroundColor: this.props.color, opacity: this.props.course.image_download_url ? 0.8 : 1 }]} />
              </View>
            </View>
            <View style={styles.row}>
              <View style={styles.rowContent}>
                <Text style={styles.primaryText} testID='course-settings.name-lbl'>
                  {i18n('Name')}
                </Text>
                <TextInput
                  value={this.state.name}
                  style={styles.actionableText}
                  onChangeText={(text) => this.setState({ name: text })}
                  testID='course-settings.name-input-textbox'
                />
              </View>
            </View>
            <View style={styles.separator}/>
            <TouchableHighlight
              underlayColor="#eee"
              onPress={this._togglePicker}
              testID='course-settings.toggle-home-picker'
            >
              <View style={styles.row}>
                <View style={styles.rowContent}>
                  <Text style={styles.primaryText} testID='course-settings.set-home-lbl'>
                    {i18n(`Set 'Home' to...`)}
                  </Text>
                  <Text
                    style={[styles.actionableText, pickerDetailStyle]}
                    testID='course-settings.home-page-lbl'>
                    { DISPLAY_NAMES.get(this.state.home) }
                  </Text>
                </View>
              </View>
            </TouchableHighlight>
            <View style={styles.separator}/>
            { this.state.showingPicker &&
              <PickerIOS
                style={styles.picker}
                selectedValue={this.state.home}
                onValueChange={(home) => this.setState({ home: home })}
                testID='course-settings.home-picker'>
                {Array.from(DISPLAY_NAMES.keys()).map((key) => (
                  <PickerItemIOS
                    key={key}
                    value={key}
                    label={DISPLAY_NAMES.get(key)}
                  />
                ))}
              </PickerIOS>
            }
            <View
              style={[styles.fakePickerDrawer, { height: this.state.showingPicker ? 0 : 150 }]} />
          </KeyboardAwareScrollView>
        </View>
      </Screen>
    )
  }
}

let connected = connect(mapStateToProps, CourseSettingsActions)(CourseSettings)
export default (connected: CourseSettings)

const styles = StyleSheet.create({
  scrollView: {
    flex: 1,
  },
  header: {
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
    paddingTop: 20,
    height: 235,
  },
  headerContent: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  row: {
    flex: 1,
    flexDirection: 'column',
    height: 54,
    paddingHorizontal: 16,
    justifyContent: 'center',
  },
  rowContent: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  primaryText: {
    flex: 2,
    fontWeight: '600',
    fontSize: 16,
    color: colors.darkText,
    lineHeight: 54,
  },
  actionableText: {
    flex: 3,
    color: colors.darkText,
    fontSize: 16,
    textAlign: 'right',
    lineHeight: 54,
  },
  separator: {
    backgroundColor: '#C7CDD1',
    height: StyleSheet.hairlineWidth,
  },
  picker: {
    flex: 1,
  },
  fakePickerDrawer: {
    backgroundColor: 'white',
    flex: 0,
  },
})
