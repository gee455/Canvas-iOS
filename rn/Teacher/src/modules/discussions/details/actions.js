//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import { createAction } from 'redux-actions'
import canvas from '../../../canvas-api'

export let Actions = (api: CanvasApi): * => ({
  refreshDiscussionEntries: createAction('discussionDetailEntries.refresh', (context: string, contextID: string, discussionID: string, includeNewEntries: boolean) => {
    return {
      promise: Promise.all([
        api.getAllDiscussionEntries(context, contextID, discussionID, includeNewEntries),
        api.getDiscussion(context, contextID, discussionID),
      ]).then(([view, discussion]) => {
        if (discussion.data.assignment_id && context !== 'groups') {
          return Promise.all([
            Promise.resolve(view),
            Promise.resolve(discussion),
            api.getAssignment(contextID, discussion.data.assignment_id),
          ])
        }
        return Promise.resolve([view, discussion])
      }),
      context,
      contextID,
      discussionID,
      handlesError: true,
    }
  }),
  refreshSingleDiscussion: createAction('discussions.details.refreshSingle', (context: string, contextID: string, discussionID: string) => {
    return {
      promise: api.getDiscussion(context, contextID, discussionID),
      discussionID,
    }
  }),
  deleteDiscussionEntry: createAction('discussionDetail.delete-entry', (context: string, contextID: string, discussionID: string, entryID: string, localIndexPath: number[]) => {
    return {
      promise: api.deleteDiscussionEntry(context, contextID, discussionID, entryID),
      context,
      contextID,
      discussionID,
      entryID,
      localIndexPath,
    }
  }),
  createEntry: createAction('discussions.details.createEntry', (context: string, contextID: string, discussionID: string, entryID: string = '', parameters: CreateEntryParameters, indexPath: number[], lastReplyAt: string) => {
    return {
      promise: api.createEntry(context, contextID, discussionID, entryID, parameters),
      context,
      contextID,
      discussionID,
      entryID,
      indexPath,
      lastReplyAt,
    }
  }),
  markEntryAsRead: createAction('discussions.details.markEntryAsRead', (context: string, contextID: string, discussionID: string, entryID: string) => {
    return {
      promise: api.markEntryAsRead(context, contextID, discussionID, entryID),
      discussionID,
      entryID,
    }
  }),
  markAllAsRead: createAction('discussions.details.markAllAsRead', (context: string, contextID: string, discussionID: string, oldUnreadCount: number) => {
    return {
      promise: api.markAllAsRead(context, contextID, discussionID),
      discussionID,
      oldUnreadCount,
    }
  }),
  editEntry: createAction('discussions.details.editEntry', (context: string, contextID: string, discussionID: string, entryID: string, parameters: CreateEntryParameters, indexPath: number[]) => {
    return {
      promise: api.editEntry(context, contextID, discussionID, entryID, parameters),
      context,
      contextID,
      discussionID,
      entryID,
      indexPath,
    }
  }),
  deletePendingReplies: createAction('discussions.details.deletePendingReplies', (discussionID) => ({
    discussionID,
  })),
})

export default (Actions(canvas): *)
