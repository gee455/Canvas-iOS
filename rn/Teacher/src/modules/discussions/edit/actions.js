//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import { createAction } from 'redux-actions'
import canvas from '../../../canvas-api'

export let Actions = (api: CanvasApi): * => ({
  createDiscussion: createAction('discussions.edit.create', (context: CanvasContext, contextID: string, params: CreateDiscussionParameters) => {
    return {
      promise: api.createDiscussion(context, contextID, params),
      params,
      handlesError: true,
      context,
      contextID,
    }
  }),
  updateDiscussion: createAction('discussions.edit.update', (context: CanvasContext, contextID: string, params: UpdateDiscussionParameters) => {
    return {
      promise: api.updateDiscussion(context, contextID, params),
      params,
      handlesError: true,
      context,
      contextID,
    }
  }),
  deleteDiscussion: createAction('discussions.edit.delete', (context: CanvasContext, contextID: string, discussionID: string) => {
    return {
      promise: api.deleteDiscussion(context, contextID, discussionID),
      discussionID,
      context,
      contextID,
    }
  }),
  subscribeDiscussion: createAction('discussions.edit.subscribe', (context: CanvasContext, contextID: string, discussionID: string, subscribed: boolean) => {
    return {
      promise: api.subscribeDiscussion(context, contextID, discussionID, subscribed),
      discussionID,
      context,
      contextID,
      subscribed,
    }
  }),
  deletePendingNewDiscussion: createAction('discussions.edit.deletePendingNew', (context: CanvasContext, contextID: string) => ({
    context,
    contextID,
  })),
})

export default (Actions(canvas): *)
