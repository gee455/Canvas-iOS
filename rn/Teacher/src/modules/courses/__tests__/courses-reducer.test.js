//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import { CoursesActions } from '../actions'
import { CourseSettingsActions } from '../settings/actions'
import { EnrollmentsActions } from '../../enrollments/actions'
import { courses as coursesReducer } from '../courses-reducer'
import { apiResponse, apiError } from '../../../../test/helpers/apiMock'
import { testAsyncReducer } from '../../../../test/helpers/async'
import * as templates from '../../../__templates__'

describe('courses refresher', () => {
  it('should capture courses from response', async () => {
    const course = templates.course()
    const courses = [course]
    const customColors = templates.customColors()

    let action = CoursesActions({
      getCourses: apiResponse(courses),
      getCustomColors: apiResponse(customColors),
    }).refreshCourses()

    let state = await testAsyncReducer(coursesReducer, action)

    const expected: CourseState = {
      color: '#fff',
      course: course,
      pending: 0,
      tabs: {
        pending: 0,
        tabs: [],
      },
      assignmentGroups: {
        pending: 0,
        refs: [],
      },
      attendanceTool: {
        pending: 0,
      },
      enrollments: {
        pending: 0,
        refs: [],
      },
      quizzes: {
        pending: 0,
        refs: [],
      },
      discussions: {
        pending: 0,
        refs: [],
      },
      announcements: {
        pending: 0,
        refs: [],
      },
      groups: {
        pending: 0,
        refs: [],
      },
      enabledFeatures: [],
      gradingPeriods: {
        pending: 0,
        refs: [],
      },
      permissions: null,
    }
    expect(state).toEqual([{}, {
      [course.id]: expected,
    }])
  })

  it('puts in all courses', async () => {
    const course = templates.course()
    const nonTeacherCourse = { ...templates.course({ id: 991 }), enrollments: [] }
    const courses = [course, nonTeacherCourse]
    const customColors = templates.customColors()

    let action = CoursesActions({
      getCourses: apiResponse(courses),
      getCustomColors: apiResponse(customColors),
    }).refreshCourses()

    let state = await testAsyncReducer(coursesReducer, action)
    expect(state).toMatchSnapshot()
  })

  it('refresh courses with error', async () => {
    let action = CoursesActions({ getCourses: apiError({ message: 'no courses' }), getCustomColors: apiError({ message: 'no courses' }) }).refreshCourses()
    let state = await testAsyncReducer(coursesReducer, action)

    // the courses store doesn't track errors or pending
    expect(state).toEqual([{}, {}])
  })
})

describe('update custom color', () => {
  it('should change the color on pending', async () => {
    let action = CoursesActions({
      updateCourseColor: apiResponse({ hexcode: '#fff' }),
    }).updateCourseColor('1', '#fff')

    let defaultState = {
      '1': {
        color: '#333',
      },
    }

    let state = await testAsyncReducer(coursesReducer, action, defaultState)
    expect(state).toMatchObject([
      {
        '1': {
          color: '#fff',
          oldColor: '#333',
        },
      },
      {
        '1': {
          color: '#fff',
        },
      },
    ])
  })

  it('reverts the color when there is an error', async () => {
    let action = CoursesActions({
      updateCourseColor: apiError({ message: 'There was an error yo' }),
    }).updateCourseColor('1', '#fff')

    let defaultState = {
      '1': {
        color: '#333',
      },
    }

    let state = await testAsyncReducer(coursesReducer, action, defaultState)
    expect(state).toMatchObject([
      {
        '1': {
          color: '#fff',
          oldColor: '#333',
        },
      },
      {
        '1': {
          color: '#333',
        },
      },
    ])
  })
})

describe('update course', () => {
  let course
  let newCourse
  let defaultState

  beforeEach(() => {
    course = templates.course({
      id: '1',
      name: 'Old Name',
      default_view: 'wiki',
    })

    newCourse = {
      ...course,
      name: 'New Name',
      default_view: 'feed',
    }

    defaultState = {
      '1': {
        error: 'try again',
        course,
      },
    }
  })

  it('should update the course state', async () => {
    let api = {
      updateCourse: apiResponse({ ...newCourse }),
    }
    let action = CourseSettingsActions(api).updateCourse(newCourse, course)

    let state = await testAsyncReducer(coursesReducer, action, defaultState)

    expect(state).toMatchObject([
      {
        '1': {
          pending: 1,
          course: {
            name: 'New Name',
            default_view: 'feed',
          },
        },
      },
      {
        '1': {
          pending: 0,
          course: {
            name: 'New Name',
            default_view: 'feed',
          },
          error: null,
        },
      },
    ])
  })

  it('should update the course state with original name', async () => {
    let api = {
      updateCourse: apiResponse({ name: newCourse.name }),
    }
    let action = CourseSettingsActions(api).updateCourse(newCourse, course)

    let state = await testAsyncReducer(coursesReducer, action, defaultState)

    expect(state).toMatchObject([
      {
        '1': {
          pending: 1,
          course: {
            name: 'New Name',
            default_view: 'feed',
          },
        },
      },
      {
        '1': {
          pending: 0,
          course: {
            name: 'New Name',
            original_name: 'New Name',
            default_view: 'feed',
          },
          error: null,
        },
      },
    ])
  })

  it('should update the course state with differing name and original name', async () => {
    let api = {
      updateCourse: apiResponse({ original_name: newCourse.name, name: 'nickname' }),
    }
    let action = CourseSettingsActions(api).updateCourse(newCourse, course)

    let state = await testAsyncReducer(coursesReducer, action, defaultState)

    expect(state).toMatchObject([
      {
        '1': {
          pending: 1,
          course: {
            name: 'New Name',
            default_view: 'feed',
          },
        },
      },
      {
        '1': {
          pending: 0,
          course: {
            name: 'nickname',
            original_name: 'New Name',
            default_view: 'feed',
          },
          error: null,
        },
      },
    ])
  })

  it('should revert the course state on rejected', async () => {
    let api = {
      updateCourse: apiError({ message: 'error' }),
    }
    let action = CourseSettingsActions(api).updateCourse(newCourse, course)

    let state = await testAsyncReducer(coursesReducer, action, defaultState)

    expect(state).toMatchObject([
      {
        '1': {
          pending: 1,
          course: {
            name: 'New Name',
            default_view: 'feed',
          },
        },
      },
      {
        '1': {
          pending: 0,
          course: {
            name: 'Old Name',
            default_view: 'wiki',
          },
          error: 'error',
        },
      },
    ])
  })
})

describe('update course nickname', () => {
  let course
  let nickname = 'nickname'
  let defaultState

  beforeEach(() => {
    course = templates.course({
      id: '1',
      name: 'Old Name',
      default_view: 'wiki',
    })

    defaultState = {
      '1': {
        error: 'try again',
        course,
        pending: 0,
      },
    }
  })

  it('should update the course nickname', async () => {
    let api = {
      updateCourseNickname: apiResponse({ nickname, name: course.name }),
    }
    let action = CoursesActions(api).updateCourseNickname(course, nickname)
    let state = await testAsyncReducer(coursesReducer, action, defaultState)

    expect(state).toMatchObject([
      {
        '1': {
          pending: 1,
          course: {
            name: course.name,
          },
        },
      },
      {
        '1': {
          pending: 0,
          course: {
            name: nickname,
            original_name: course.name,
          },
          error: null,
        },
      },
    ])
  })
})

describe('getCourseEnabledFeature', () => {
  it('should set the store to pending when the action is called', () => {
    let action = {
      type: CoursesActions().getCourseEnabledFeatures.toString(),
      pending: true,
      payload: {
        courseID: '1',
      },
    }

    let newState = coursesReducer({}, action)
    expect(newState).toMatchObject({
      '1': {
        pending: 1,
      },
    })
  })

  it('should set the store not to pending whent the action is rejected', () => {
    let action = {
      type: CoursesActions().getCourseEnabledFeatures.toString(),
      error: true,
      payload: {
        courseID: '1',
      },
    }

    let state = {
      '1': {
        pending: 1,
      },
    }
    let newState = coursesReducer(state, action)
    expect(newState).toMatchObject({
      '1': {
        pending: 0,
      },
    })
  })

  it('should set the enabledFeatures on the course when it is successful', () => {
    let action = {
      type: CoursesActions().getCourseEnabledFeatures.toString(),
      payload: {
        courseID: '1',
        result: {
          data: ['anonymous_grading'],
        },
      },
    }

    let state = {
      '1': {
        pending: 1,
      },
    }
    let newState = coursesReducer(state, action)
    expect(newState).toMatchObject({
      '1': {
        pending: 0,
        enabledFeatures: ['anonymous_grading'],
      },
    })
  })
})

describe('getCoursePermissions', () => {
  it('should set the permissions on the course', () => {
    let action = {
      type: CoursesActions().getCoursePermissions.toString(),
      payload: {
        courseID: '1',
        result: {
          data: { send_messages: false },
        },
      },
    }

    let state = {
      '1': {},
    }
    let newState = coursesReducer(state, action)
    expect(newState).toMatchObject({
      '1': {
        permissions: { send_messages: false },
      },
    })
  })
})

describe('refresh single course', () => {
  it('no existing permissions', () => {
    let action = {
      type: CoursesActions().refreshCourse.toString(),
      payload: {
        result: {
          data: {
            permissions: {
              create_announcement: true,
              create_discussion_topic: true,
            }
            ,
          },
        },
        context: 'courses',
        contextID: '1',
      },
    }

    let state = {}
    let newState = coursesReducer(state, action)
    expect(newState).toMatchObject(
      {
        '1': {
          'announcements': { 'pending': 0, 'refs': [] },
          'assignmentGroups': { 'pending': 0, 'refs': [] },
          'attendanceTool': { 'pending': 0 },
          'color': '#FFFFFF00',
          'course': {},
          'discussions': { 'pending': 0, 'refs': [] },
          'enabledFeatures': [],
          'enrollments': { 'pending': 0, 'refs': [] },
          'error': null,
          'gradingPeriods': { 'pending': 0, 'refs': [] },
          'groups': { 'pending': 0, 'refs': [] },
          'oldColor': '#FFFFFF00',
          'pending': 0,
          'permissions': {},
          'quizzes': { 'pending': 0, 'refs': [] },
          'tabs': { 'pending': 0, 'tabs': [] },
        },
        'undefined': { 'permissions': { 'create_announcement': true, 'create_discussion_topic': true } },
      }
    )
  })

  it('existing permissions', () => {
    let action = {
      type: CoursesActions().refreshCourse.toString(),
      payload: {
        result: {
          data: {
            permissions: {
              create_announcement: false,
              create_discussion_topic: true,
            }
            ,
          },
        },
        context: 'courses',
        contextID: '1',
      },
    }

    let state = {
      '1': {
        'announcements': { 'pending': 0, 'refs': [] },
        'assignmentGroups': { 'pending': 0, 'refs': [] },
        'attendanceTool': { 'pending': 0 },
        'color': '#FFFFFF00',
        'course': {},
        'discussions': { 'pending': 0, 'refs': [] },
        'enabledFeatures': [],
        'enrollments': { 'pending': 0, 'refs': [] },
        'error': null,
        'gradingPeriods': { 'pending': 0, 'refs': [] },
        'groups': { 'pending': 0, 'refs': [] },
        'oldColor': '#FFFFFF00',
        'pending': 0,
        'permissions': {},
        'quizzes': { 'pending': 0, 'refs': [] },
        'tabs': { 'pending': 0, 'tabs': [] },
      },
      'undefined': { 'permissions': { 'create_announcement': false, 'create_discussion_topic': true } },
    }
    let newState = coursesReducer(state, action)
    expect(newState).toMatchObject(
      {
        '1': {
          'announcements': { 'pending': 0, 'refs': [] },
          'assignmentGroups': { 'pending': 0, 'refs': [] },
          'attendanceTool': { 'pending': 0 },
          'color': '#FFFFFF00',
          'course': {},
          'discussions': { 'pending': 0, 'refs': [] },
          'enabledFeatures': [],
          'enrollments': { 'pending': 0, 'refs': [] },
          'error': null,
          'gradingPeriods': { 'pending': 0, 'refs': [] },
          'groups': { 'pending': 0, 'refs': [] },
          'oldColor': '#FFFFFF00',
          'pending': 0,
          'permissions': {},
          'quizzes': { 'pending': 0, 'refs': [] },
          'tabs': { 'pending': 0, 'tabs': [] },
        },
        'undefined': { 'permissions': { 'create_announcement': false, 'create_discussion_topic': true } },
      }
    )
  })
})

describe('refresh users courses', () => {
  it('should add the refs to the right course', () => {
    let action = {
      type: EnrollmentsActions().refreshUserEnrollments.toString(),
      payload: {
        result: {
          data: [
            templates.enrollment({ id: '1', course_id: '1' }),
            templates.enrollment({ id: '2', course_id: '2' }),
          ],
        },
      },
    }

    let state = {
      '1': {},
      '2': {},
    }

    let newState = coursesReducer(state, action)
    expect(newState).toMatchObject({
      '1': {
        enrollments: {
          refs: ['1'],
        },
      },
      '2': {
        enrollments: {
          refs: ['2'],
        },
      },
    })
  })

  it('should create initial course state if no course is present', () => {
    let action = {
      type: EnrollmentsActions().refreshUserEnrollments.toString(),
      payload: {
        result: {
          data: [
            templates.enrollment({ id: '1', course_id: '1' }),
          ],
        },
      },
    }

    let state = {}
    let newState = coursesReducer(state, action)
    expect(newState).toMatchObject({
      '1': {
        enrollments: {
          refs: ['1'],
        },
      },
    })
    expect(newState['1'].color).not.toBeUndefined()
  })
})
