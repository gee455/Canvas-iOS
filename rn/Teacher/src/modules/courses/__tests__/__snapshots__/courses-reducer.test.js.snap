// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`courses refresher puts in all courses 1`] = `
Array [
  Object {},
  Object {
    "1": Object {
      "announcements": Object {
        "pending": 0,
        "refs": Array [],
      },
      "assignmentGroups": Object {
        "pending": 0,
        "refs": Array [],
      },
      "attendanceTool": Object {
        "pending": 0,
      },
      "color": "#fff",
      "course": Object {
        "account_id": "1",
        "course_code": "rn 101",
        "default_view": "wiki",
        "end_at": "2030-02-08T23:13:39Z",
        "enrollments": Array [
          Object {
            "enrollment_state": "active",
            "role": "TeacherEnrollment",
            "role_id": "1",
            "type": "teacher",
            "user_id": "1",
          },
        ],
        "id": "1",
        "image_download_url": "https://farm3.staticflickr.com/2926/14690771011_945f91045a.jpg",
        "is_favorite": true,
        "name": "Learn React Native",
        "permissions": Object {
          "viewAllGrades": true,
          "viewAnalytics": true,
        },
        "short_name": "rn",
        "term": Object {
          "created_at": "2018-02-08T23:13:39Z",
          "end_at": "2030-02-08T23:13:39Z",
          "id": "1",
          "name": "Default Term",
          "start_at": "2018-02-08T23:13:39Z",
          "workflow_state": "available",
        },
        "workflow_state": "available",
      },
      "discussions": Object {
        "pending": 0,
        "refs": Array [],
      },
      "enabledFeatures": Array [],
      "enrollments": Object {
        "pending": 0,
        "refs": Array [],
      },
      "gradingPeriods": Object {
        "pending": 0,
        "refs": Array [],
      },
      "groups": Object {
        "pending": 0,
        "refs": Array [],
      },
      "pending": 0,
      "permissions": null,
      "quizzes": Object {
        "pending": 0,
        "refs": Array [],
      },
      "tabs": Object {
        "pending": 0,
        "tabs": Array [],
      },
    },
    "991": Object {
      "announcements": Object {
        "pending": 0,
        "refs": Array [],
      },
      "assignmentGroups": Object {
        "pending": 0,
        "refs": Array [],
      },
      "attendanceTool": Object {
        "pending": 0,
      },
      "color": "#aaa",
      "course": Object {
        "account_id": "1",
        "course_code": "rn 101",
        "default_view": "wiki",
        "end_at": "2030-02-08T23:13:39Z",
        "enrollments": Array [],
        "id": 991,
        "image_download_url": "https://farm3.staticflickr.com/2926/14690771011_945f91045a.jpg",
        "is_favorite": true,
        "name": "Learn React Native",
        "permissions": Object {
          "viewAllGrades": true,
          "viewAnalytics": true,
        },
        "short_name": "rn",
        "term": Object {
          "created_at": "2018-02-08T23:13:39Z",
          "end_at": "2030-02-08T23:13:39Z",
          "id": "1",
          "name": "Default Term",
          "start_at": "2018-02-08T23:13:39Z",
          "workflow_state": "available",
        },
        "workflow_state": "available",
      },
      "discussions": Object {
        "pending": 0,
        "refs": Array [],
      },
      "enabledFeatures": Array [],
      "enrollments": Object {
        "pending": 0,
        "refs": Array [],
      },
      "gradingPeriods": Object {
        "pending": 0,
        "refs": Array [],
      },
      "groups": Object {
        "pending": 0,
        "refs": Array [],
      },
      "pending": 0,
      "permissions": null,
      "quizzes": Object {
        "pending": 0,
        "refs": Array [],
      },
      "tabs": Object {
        "pending": 0,
        "tabs": Array [],
      },
    },
  },
]
`;
